### PR TITLE
Insight api endpoints

### PR DESCRIPTION
**Pull request by samparsky**
This is the implementation of insight api endpoints for v5.3.04beta. It uses postgres database and isn't enabled if the node doesn't have PG enabled.

It implements all the endpoints found here
https://github.com/bitpay/insight-api#api-http-endpoints

### CODE REVIEW

**Comment by chappjc** [comment]
Overall looks like a really good start.  Needs a rebase, cleanup, and more docs.  Also, code reuse is not always bad, and since we're supporting a separate API it's inevitable, but it seems like we might be able to reduce it some.  For instance, the middleware is still really really close to the existing middleware and I'm not sure that benefits from a unique impl.

**Comment by samparsky** [comment]
Okay then, will make the middleware implementation a package for code reuse. And squash the commits.

**Comment by chappjc** [comment]
Thanks.  I think it's worth the effort

**Comment by chappjc** [comment]
Please also address the linter issues: https://travis-ci.org/dcrdata/dcrdata/jobs/342204160#L708-L737

File: apiroutes.go [hunk 1]
```diff
@@ -60,22 +62,36 @@ type APIDataSource interface {
...
 // dcrdata application context used by all route handlers
 type appContext struct {
-	nodeClient *rpcclient.Client
-	BlockData  APIDataSource
-	Status     apitypes.Status
-	statusMtx  sync.RWMutex
-	JSONIndent string
+	nodeClient     *rpcclient.Client
+	BlockData      APIDataSource
+	ExplorerSource explorerDataSource
```
Comment (chappjc) [inline apiroutes.go]: is this used?

File: apiroutes.go [hunk 2]
```diff
@@ -86,6 +102,8 @@ func newContext(client *rpcclient.Client, blockData APIDataSource, JSONIndent st
 		},
 		JSONIndent: JSONIndent,
 	}
+
+	return &newAppContext
```
Comment (chappjc) [inline apiroutes.go]: why this change?

File: db/dcrpg/insightapi.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
+package dcrpg
```
Comment (chappjc) [inline db/dcrpg/insightapi.go]: Add copyright, as in other files.

File: db/dcrpg/insightapi.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+import (
+	"errors"
+	"github.com/dcrdata/dcrdata/db/dbtypes"
```
Comment (chappjc) [inline db/dcrpg/insightapi.go]: blank line between std library imports and github

File: db/dcrpg/insightapi.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+	}
+	txs, err := pgb.client.SearchRawTransactionsVerbose(address, skip, count, true, true, nil)
+	if err != nil {
+		log.Warnf("GetAddressTransactions failed for address %s: %v", addr, err)
+		return nil
+	}
+	return txs
+}
+
+// GetTransactionHex returns hex representation of
```
Comment (chappjc) [inline db/dcrpg/insightapi.go]: please wrap to 80 chars

File: db/dcrpg/insightapi.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+	if err != nil {
+		log.Errorf("GetRawTransactionVerbose failed for: %v", err)
+		return ""
+	}
+
+	return txraw.Hex
+}
+
+// GetBlockVerboseByHash returns the verbose of a block by the block hash
```
Comment (chappjc) [inline db/dcrpg/insightapi.go]: "the verbose of a block" doesnt' make much sense.  If you're not sure, just say it returns a dcrjson.GetBlockVerboseResult

File: db/dcrpg/insightapi.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+	blockTransactions.Tx = make([]string, len(blockVerbose.Tx))
+	copy(blockTransactions.Tx, blockVerbose.Tx)
+
+	blockTransactions.STx = make([]string, len(blockVerbose.STx))
+	copy(blockTransactions.STx, blockVerbose.STx)
+
+	return blockTransactions
+}
+
+// GetBlockHash returns the block hash by
```
Comment (chappjc) [inline db/dcrpg/insightapi.go]: wrapping

File: dcrdataapi/apitypes.go [hunk 8]
```diff
@@ -352,3 +353,16 @@ type MempoolTicketDetails struct {
 // TicketsDetails is an array of pointers of TicketDetails used in
 // MempoolTicketDetails
 type TicketsDetails []*TicketDetails
+
+// insight api_version
```
Comment (chappjc) [inline dcrdataapi/apitypes.go]: begin comment with "AddressTxnOutput"

File: dcrdataapi/insightapitypes.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
+package dcrdataapi
```
Comment (chappjc) [inline dcrdataapi/insightapitypes.go]: copyright

File: dcrdataapi/insightapitypes.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	"github.com/decred/dcrd/dcrjson"
+	"github.com/decred/dcrd/dcrutil"
+)
+
+// AddressTxShort is a subset of AddressTxRaw with just the basic tx details
```
Comment (chappjc) [inline dcrdataapi/insightapitypes.go]: remove irrelevant comment and doc each type below.

File: dcrdataapi/insightapitypes.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+type InsightAddressInfo struct {
+	Address          string         `json:"addrStr,omitempty"`
+	Limit            int64          `json:"limit,omitemtpy"`
+	Offset           int64          `json:"offset,omitempty"`
+	TransactionsID   []string       `json:"transactions,omitempty"`
+	NumFundingTxns   int64          `json:"numFundingTxns,omitempty"`
+	NumSpendingTxns  int64          `json:"numSpendingTxns,omitempty"`
+	KnownFundingTxns int64          `json:"knownFundingTxns,omitempty"`
+	NumUnconfirmed   int64          `json:"numUnconfirmed,omitempty"`
+	TotalReceived    dcrutil.Amount `json:"totalReceived"`
```
Comment (chappjc) [inline dcrdataapi/insightapitypes.go]: be sure `dcrutil.Amount` marshals to JSON as intended

File: explorer/explorertypes.go [hunk 12]
```diff
@@ -211,6 +211,20 @@ type ChainParams struct {
...
+// type AddressTxnOutput struct {
+// 	Address       string
+// 	TxnID         string
+// 	Vout          int
+// 	ScriptPubKey  string
+// 	Height        int64
+// 	Amount        float64
+// 	Atoms         float64
+// 	Confirmations uint64
+// }
```
Comment (chappjc) [inline explorer/explorertypes.go]: remove

File: insightapi/apimiddleware.go [hunk 13]
```diff
@@ -0,0 +1,211 @@
+package insightapi
```
Comment (chappjc) [inline insightapi/apimiddleware.go]: as mentioned in comments, please create middlware package since this is a ton of copy/paste from https://github.com/dcrdata/dcrdata/blob/master/apimiddleware.go

File: insightapi/apiroutes.go [hunk 14]
```diff
@@ -0,0 +1,482 @@
...
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+	"sync"
+	"github.com/dcrdata/dcrdata/db/dbtypes"
+	apitypes "github.com/dcrdata/dcrdata/dcrdataapi"
+	"fmt"
+	"github.com/dcrdata/dcrdata/db/dcrpg"
```
Comment (chappjc) [inline insightapi/apiroutes.go]: please separate standard library imports

File: rpcutils/rpcclient.go [hunk 15]
```diff
@@ -220,3 +220,34 @@ func GetBlock(ind int64, client *rpcclient.Client) (*dcrutil.Block, *chainhash.H
 
 	return block, blockhash, nil
 }
+
+// GetTransaction get a transaction with a given hash
```
Comment (chappjc) [inline rpcutils/rpcclient.go]: // GetTransactionVerboseByID...

Comment (chappjc) [reply rpcutils/rpcclient.go]: but this is a confusing name since it's taking a hash.  Let's call it `GetTransactionVerbose` unless that already exists

File: rpcutils/rpcclient.go [hunk 16]
```diff
@@ -220,3 +220,34 @@ func GetBlock(ind int64, client *rpcclient.Client) (*dcrutil.Block, *chainhash.H
...
+}
+
+func SearchRawTransaction(client *rpcclient.Client, address string) ([]*dcrjson.SearchRawTransactionsResult, error) {
+	addr, err := dcrutil.DecodeAddress(address)
+	if err != nil {
+		log.Infof("Invalid address %s: %v", address, err)
+		return nil, err
+	}
+	//change the 1000 000 number demo for now
+	txs, err := client.SearchRawTransactionsVerbose(addr, 0, 20,
```
Comment (chappjc) [inline rpcutils/rpcclient.go]: 20?

File: db/dcrpg/insightapi.go [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+		}
+	}
+
+	tx.Amount = tx.Atoms * 100000000
+	return tx
+}
+
+// GetAddressUTXO returns the utxo of an address
+func (db *ChainDB) GetAddressUTXO(address string) []*apitypes.AddressTxnOutput {
+	txs, err := rpcutils.SearchRawTransaction(db.client, address)
```
Comment (chappjc) [inline db/dcrpg/insightapi.go]: How do we deal with the total and the count/skip limits?

File: rpcutils/rpcclient.go [hunk 18]
```diff
@@ -220,3 +220,34 @@ func GetBlock(ind int64, client *rpcclient.Client) (*dcrutil.Block, *chainhash.H
...
+		true, true, nil)
+	if err != nil {
+		log.Warnf("GetAddressUTXO failed for address %s: %v", addr, err)
```
Comment (chappjc) [inline rpcutils/rpcclient.go]: wrong message

File: db/dcrpg/insightapi.go (hunk 17 above)
Comment (chappjc) [reply db/dcrpg/insightapi.go]: We should get the needed data from postgresql here, not RPC.  You'll have to look up the vouts for the input address, then get the rest of the info needed by whatever insight endpoint this is for.

File: db/dcrsqlite/apisource.go [hunk 19]
```diff
@@ -1117,6 +1178,26 @@ func (db *wiredDB) GetExplorerTx(txid string) *explorer.TxInfo {
 	return tx
 }
 
+func (db *wiredDB) GetAddressUTXO(address string) []*apitypes.AddressTxnOutput {
```
Comment (chappjc) [inline db/dcrsqlite/apisource.go]: why implement this for `wiredDB` too?  Do we intend for this to be used in lite mode as well as full with PG?  If so, it's time to write an function that repeatedly calls `SearchRawTransactionsVerbose` until it gets all the transactions, although we've found that is very demanding on dcrd.

File: db/dcrsqlite/sqlite.go [hunk 20]
```diff
@@ -511,6 +516,33 @@ func (db *DB) RetrieveSDiffRange(ind0, ind1 int64) ([]float64, error) {
...
 
+func (db *DB) RetrieveBlockSummaryByTimeRange(minTime, maxTime int64, limit int) ([]apitypes.BlockDataBasic, error) {
+	blocks := make([]apitypes.BlockDataBasic, 0, limit)
+
+	stmt, err := db.Prepare(db.getBlockByTimeRangeSQL)
+	if err != nil {
+		return nil, err
+	}
+
+	rows, err := stmt.Query(minTime, maxTime, limit)
```
Comment (chappjc) [inline db/dcrsqlite/sqlite.go]: `err` unchecked

File: db/dcrpg/internal/blockstmts.go [hunk 21]
```diff
@@ -90,7 +92,12 @@ const (
 	VALUES ($1, $2, $3, $4)
 	ON CONFLICT (this_hash) DO NOTHING;`
 
-	UpdateBlockNext = `UPDATE block_chain set next_hash = $2 WHERE block_db_id = $1;`
+	SelectBlockHashByHeight = `select hash from blocks where height = $1`
+	SelectBlockHeightByHash = `select height from blocks where hash = $1`
+
+
+
+UpdateBlockNext = `UPDATE block_chain set next_hash = $2 WHERE block_db_id = $1;`
```
Comment (chappjc) [inline db/dcrpg/internal/blockstmts.go]: formatting

File: db/dcrpg/pgblockchain.go [hunk 22]
```diff
@@ -32,13 +33,14 @@ var (
...
-	lastBlock          map[chainhash.Hash]uint64
-	addressCounts      *addressCounter
+	db            *sql.DB
+	chainParams   *chaincfg.Params
+	devAddress    string
+	dupChecks     bool
+	bestBlock     int64
+	lastBlock     map[chainhash.Hash]uint64
+	addressCounts *addressCounter
+	client        *rpcclient.Client
```
Comment (chappjc) [inline db/dcrpg/pgblockchain.go]: I do not like adding rpcclient.Client to `ChainDB`.  I'd prefer a new type that embeds `ChainDB` and adds the client.  Thus, all the new ChainDB methods you added would then belong to that new type instead.

File: db/dcrpg/pgblockchain.go [hunk 23]
```diff
@@ -358,8 +373,10 @@ func (pgb *ChainDB) AddressHistory(address string, N, offset int64) ([]*dbtypes.
 			}
 		} else {
 			var numSpent, numUnspent, totalSpent, totalUnspent int64
+
```
Comment (chappjc) [inline db/dcrpg/pgblockchain.go]: remove this blank line please

File: db/dcrsqlite/apisource.go [hunk 24]
```diff
@@ -654,6 +654,15 @@ func (db *wiredDB) GetPoolByHash(hash string) ([]string, error) {
 	return hss, nil
 }
 
+
+func (db *wiredDB) GetBlockSummaryTimeRange(min, max int64, limit int) []apitypes.BlockDataBasic {
```
Comment (chappjc) [inline db/dcrsqlite/apisource.go]: docs

File: db/dcrsqlite/apisource.go [hunk 25]
```diff
@@ -737,6 +746,36 @@ func (db *wiredDB) GetMempoolSSTxDetails(N int) *apitypes.MempoolTicketDetails {
 	return &mpTicketDetails
 }
 
+// GetAddressTransactions returns an apitypes.Address Object with at most the
```
Comment (chappjc) [inline db/dcrsqlite/apisource.go]: // GetAddressTransactionsWithSkip ..

**Comment by samparsky** [comment]
Already working on implementing the changes

**Comment by samparsky** [comment]
@chappjc What about the ntfnchans.go and ntfnhandlers.go file

**Comment by chappjc** [comment]
Those can be factored into their own package (if it makes sense) in a separate PR.

Please rebase.  Thanks

**Comment by samparsky** [comment]
@chappjc The apiroutes.go file is dependent on those files it uses ntfchan struct, so refactoring apiroutes require refactoring the nfchan files.

**Comment by chappjc** [comment]
Gotcha.  Those can go into a small package of their own.  Unless there's another logical place for them where there's little risk is circular depends

**Comment by chappjc** [comment]
When ready and tested, please state ready for review.  Thanks.

**Comment by chappjc** [comment]
Looks like you didn't use `git mv` to move ntfnchans.go to notification/ntfnchans.go.  Please get the file back and use git mv

**Comment by chappjc** [comment]
On second thought, don't stress about getting ntfnchans.go to show as a move.  the diff is large enough that git doesn't want to show it that way.

**Comment by chappjc** [comment]
Please address build and lint errors: https://travis-ci.org/decred/dcrdata/jobs/352598292

**Comment by chappjc** [comment]
@samparsky Thanks.  Will test.

File: api/insight/apiroutes.go [hunk 26]
```diff
@@ -0,0 +1,393 @@
...
+	Status     apitypes.Status
+	statusMtx  sync.RWMutex
+	JSONIndent string
+}
+
+// NewInsightContext Constructor for insightApiContext
+func NewInsightContext(client *rpcclient.Client, blockData *dcrpg.ChainDBRPC, JSONIndent string) *insightApiContext {
+	conns, _ := client.GetConnectionCount()
+	nodeHeight, _ := client.GetBlockCount()
+	version := semver.NewSemver(0, 0, 1)
```
Comment (chappjc) [inline api/insight/apiroutes.go]: please set to whatever api version we're implementing

Comment (chappjc) [reply api/insight/apiroutes.go]: According to Insight docs, they are at API version 1.0

https://documentation.riada.se/insight/latest/insight-advanced-usage-guide/insight-for-developers/insight-rest-api

So 1.0.0 here.

File: api/version.go [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+
+import "fmt"
+
+type version struct {
+	Major, Minor, Patch int
+	Label               string
+	Nick                string
+}
+
+var ver = version{
```
Comment (chappjc) [inline api/version.go]: Version of the `api` package.  It's a new package, so please start at 1.0.0

File: config.go [hunk 28]
```diff
@@ -1,5 +1,4 @@
 // Copyright (c) 2016-2018 The Decred developers
-// Copyright (c) 2017 Jonathan Chappelow
```
Comment (chappjc) [inline config.go]: Do not remove copyright holders.  Only prepend.

File: main.go [hunk 29]
```diff
@@ -1,6 +1,3 @@
-// Copyright (c) 2017, Jonathan Chappelow
```
Comment (chappjc) [inline main.go]: restore the copyright.  Prepend The Decred developers.

File: api/apirouter.go [hunk 30]
```diff
@@ -1,9 +1,13 @@
-package main
+// Copyright (c) 2017, The dcrdata developers
```
Comment (chappjc) [inline api/apirouter.go]:
```
// Copyright (c) 2018, The Decred developers
// Copyright (c) 2017, The dcrdata developers
// Copyright (c) 2017, Jonathan Chappelow
// See LICENSE for details.

**Comment by chappjc** [comment]
Your rebase did not correctly perform conflict resolution.  Diff ntfnchans.go in master with yours.  Several lines have been omitted, such as:

```go
// New mempool tx chan for explorer
ntfnChans.expNewTxChan = make(chan *explorer.NewMempoolTx, expNewTxChanBuffer)
```

There are several changes you need to take care to integrate.  Also check ntfnhandlers.go

File: config.go [hunk 31]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) 2016-2018 The Decred developers
```
Comment (chappjc) [inline config.go]: What I meant was to prepend like this:

```
// Copyright (c) 2016-2018 The Decred developers
// Copyright (c) 2017 Jonathan Chappelow
// Use of this source code is governed by an ISC
// license that can be found in the LICENSE file.
```

**Comment by chappjc** [comment]
Using `jd` to diff JSON output from APIs on https://mainnet.decred.org/api and http://127.0.0.1:7777/insight-api

**`/block/{hash}`**

```sh
jd \
 <(curl -s https://mainnet.decred.org/api/block/000000000000000dc55491f53e2bef392d98eae460edd2a88147f38f7574588d) \
 <(curl -s http://127.0.0.1:7777/insight-api/block/000000000000000dc55491f53e2bef392d98eae460edd2a88147f38f7574588d)
@ ["isMainChain"]
- true
@ ["reward"]
- 21.80354156
@ ["unixtime"]
- "Mon, 19 Mar 2018 19:21:10 GMT"
```

:x:  missing data
<hr>

**`/block-index/{index}`**

```
BLKIDX=65432
jd <(curl -s https://mainnet.decred.org/api/block-index/$BLKIDX) <(curl -s http://127.0.0.1:7777/insight-api/block-index/$BLKIDX)
```

:heavy_check_mark: Same!
<hr>

**`/rawblock/{index}`**

```
curl -s http://127.0.0.1:7777/insight-api/rawblock/$BLKIDX
Implementation pending
```
:x:
<hr>

**`/rawblock/{hash}`**

```
curl -s http://127.0.0.1:7777/insight-api/rawblock/$BLKHASH
404 page not found
```
:x:
<hr>

....

please check all on https://github.com/bitpay/insight-api#api-http-endpoints
